Table of Contents

leonardoventurini opened the pull request:
## Description

I implemented a handlebars helper to render a basic Table of Contents based on git log by using the following directive in the template files: 

```
{{{ toc }}}
```

For now, the anchors follow GitHub's markdown algorithm I guess, but it is easy to implement more templating options.

### Generated Markdown Example

![image](https://user-images.githubusercontent.com/17411182/71631560-a38f4b80-2bdf-11ea-924f-7033d3d4ae94.png)

![image](https://user-images.githubusercontent.com/17411182/71631569-b43fc180-2bdf-11ea-845a-4816db3a3d1b.png)

leonardoventurini commented:
The substeps are linking to manually added subheaders... I will need to find a better way to do this. Currently, I can use my changes but other people might have a hard time using them.

So I will close the pull request and create a new pull request later on. Thanks.

Urigo commented:
thank you for the update @leonardoventurini

Urigo commented:
Looking forward to get this done.
If you need help, don't hesitate to ask!

leonardoventurini commented:
> Looking forward to get this done.
> If you need help, don't hesitate to ask!

Sure, I had some problems when running the `tortilla manual render --all` since it is recursive and the following iterations are all from within rebase, I probably need to store in the log in localStorage or something I guess... thanks

leonardoventurini commented:
The recursion problem is fixed. :tada: @Urigo

DAB0mB commented on src/git.ts:
```diff
@@ -472,6 +473,22 @@ function getRevisionIdFromObject(object: string): string {
   return git(['rev-list', '-n', '1', object]);
 }
 
+function getCWD(module?) {
```
You might wanna look at `renderer/index.ts#resolvePath()` or `utils.ts#cwd()`. Is this what you had in mind?

leonardoventurini replied:
I just figured I had to use this, and abstracted a piece of existing code into a function. I can certainly try, and perhaps replace the other call as well.

DAB0mB commented on src/manual.ts:
```diff
@@ -48,6 +50,19 @@ init();
+      .map(arr => JSON.stringify(arr, null, 4))
+      .pop();
+
+    LocalStorage.setItem('TABLE_OF_CONTENTS', log);
```
Why do you use the local storage? So you won't have to run it every time? I think that storing it in-memory somehow can be a more appropriate solution, less heavier. If not, can you add a comment in-code?

leonardoventurini replied:
@DAB0mB I use LocalStorage because tortilla runs several processes recursively to render all steps. I tried the way you said before but only worked when rendering the root template, and cause errors when rendering all of them. I also could not find a way to store in memory so it would run only for the root template and ignore other templates... it seems when we know we should get the log, the repository is well within rebasing it... and I guess we should allow users to place the Table of Contents elsewhere too. 

That is a valid concern, glad you asked, if there is a better solution please point it out to me.

DAB0mB commented:
I have added few comments. Also I cannot merge it without tests, if you have any questions about that free feel to ask.

leonardoventurini commented:
> I have added few comments. Also I cannot merge it without tests, if you have any questions about that free feel to ask.

@DAB0mB Oh, right, I forgot about the tests. I will be on that next then.

leonardoventurini commented:
> I have added few comments. Also I cannot merge it without tests, if you have any questions about that free feel to ask.

@DAB0mB There are three tests failing on master... I will see what I can do about those, but I guess I do not have much time available... but I guess you would want to check the tests... unless they are known to be failing that is.

DAB0mB commented:
@leonardoventurini If they are failing on master I will fix them, but if in the meantime you can write tests to your features that would be good. It's possible that the `git` version is causing the tests to fail.

leonardoventurini commented:
Hello @DAB0mB, I have added a test with a simple data mock for the handlebars helper. Simple stuff... how likely would it be for me to get some help setting up more complex tests? I just cannot wrap my head around the testing environment without some docs... and I am quite short on time lately. It would help me a lot if you guys took on this small feature from here onwards, less chance I might break something. Thanks.

DAB0mB commented:
@leonardoventurini There should be numerous examples in the test files. [Here's one of them](https://github.com/Urigo/tortilla/blob/3943537f27ae2986b50d2577412b22dc785d9edf/tests/manual.spec.ts#L124). You see the difference? You actually created a fake log, and you edited the local storage. That's not a good way to test things, because you don't emulate a real scenario. In practicality, what you should do is running real `git` and `tortilla` via child process utilities and render a Tortilla manual like you normally would. So it should (roughly) go like this:

```
tortilla step edit --root
bach -c 'echo "{{{ table }}}" > .tortilla/templates/root.tml'
tortilla step push -m "test step 1.1"
tortilla step push -m "test step 1.2"
tortilla step tag -m "test step 2"
tortilla step push -m "test step 2.1"
git rebase --continue
tortilla manual render --all
```

Let me know how things go.

leonardoventurini commented:
@DAB0mB I had something crazy in mind, saw all those files and thought I had to do something else entirely, turns out it is easier than I thought, I got something working, but I will add some things more until this weekend. thanks a lot.

DAB0mB commented on tests/manual.spec.ts:
```diff
@@ -208,4 +208,42 @@ describe('Manual', () => {
     });
   });
+
+  it('should render table of contents', () => {
```
Looks a lot better!

DAB0mB commented on tests/fixtures/logMock.json:
```diff
@@ -0,0 +1,16 @@
+[
```
This one can be removed, you're using the snapshot.

leonardoventurini commented:
@DAB0mB done, I added one more test to take into account the possibility of rendering the table of contents anywhere and removed the file... anything else I should do? 
there will be more pull requests in the future as our use case for the table of contents mature. next, I plan on finding a way to generate the subheaders for the substeps, since the table of contents has to point somewhere, and writing those by hand might not be ideal.
